add test for data_column parameter referring to non-tabular data (bed) and fix bug

bernt-matthias opened the pull request:
To check why https://github.com/galaxyproject/galaxy/issues/9002 is unnoticed by CI.

- [x] with https://github.com/galaxyproject/galaxy/pull/10365/commits/4c5c1350a30fc4af85143d4f4d5904ea2abf0e0b exactly the new test fails.
- [x] starting to fix using the commit(s) of https://github.com/galaxyproject/galaxy/pull/10364 (which will be closed)
- [x] check if https://github.com/galaxyproject/galaxy/blob/2777db0e46f1185e47c41049abdb61f99ac1e2e7/lib/galaxy/tools/parameters/dataset_matcher.py#L222 is necessary
- [ ] use named tuple as return

bernt-matthias committed:
convert data sets only if necessary

fixes: https://github.com/galaxyproject/galaxy/issues/9002

Co-authored-by: Marius van den Beek <m.vandenbeek@gmail.com>

bernt-matthias committed:
add test for column_param test

testing non tabular (bed) inputs

bernt-matthias committed:
differentiate direct match from un-convertable

in find_conversion_destination_for_dataset_by_extensions

mvdbeek commented on lib/galaxy/datatypes/display_applications/parameters.py:
```diff
@@ -107,8 +109,12 @@ def prepare(self, other_values, dataset_hash, user_hash, trans):
+            direct_match, target_ext, converted_dataset = data.find_conversion_destination(self.formats, converter_safe=True)
+            # TODO: Q: I guess we could skip the first if branch but code might
+            # be more readable
+            if direct_match:
```
I'd say `if not direct_match:` and then have the other 2 if statements below seems more readable ?

bernt-matthias replied:
OK.

mvdbeek commented on lib/galaxy/tools/actions/__init__.py:
```diff
@@ -88,8 +88,8 @@ def process_dataset(data, formats=None):
                     trans.sa_session.refresh(data)
-                    target_ext, converted_dataset = data.find_conversion_destination(formats)
-                    if target_ext:
+                    direct_match, target_ext, converted_dataset = data.find_conversion_destination(formats)
```
You don't need the `if not data.datatype.matches_any(formats)` statement above, you can replace that with 
```
direct_match, target_ext, converted_dataset = data.find_conversion_destination(formats)
if not direct_match:
````

mvdbeek commented on lib/galaxy/tools/parameters/dataset_matcher.py:
```diff
@@ -121,8 +121,8 @@ def valid_hda_match(self, hda, check_implicit_conversions=True):
                 return False
-            target_ext, converted_dataset = hda.find_conversion_destination(formats)
-            if target_ext:
+            direct_match, target_ext, converted_dataset = hda.find_conversion_destination(formats)
```
Same here, you can pull that up to line 119

bernt-matthias replied:
I think I covered all three comments .. please check my changes.

bernt-matthias committed:
apply suggestions from code review

Co-authored-by: Marius van den Beek <m.vandenbeek@gmail.com>

mvdbeek commented on lib/galaxy/datatypes/display_applications/parameters.py:
```diff
@@ -107,19 +109,20 @@ def prepare(self, other_values, dataset_hash, user_hash, trans):
-            elif converted_dataset and converted_dataset.state == converted_dataset.states.ERROR:
-                raise Exception("Dataset conversion failed for data parameter: %s" % self.name)
+            direct_match, target_ext, converted_dataset = data.find_conversion_destination(self.formats, converter_safe=True)
+            if direct_match:
```
```suggestion
            if not direct_match:
```

bernt-matthias replied:
You are right. I amended the change to my previous commit

mvdbeek commented on lib/galaxy/datatypes/display_applications/parameters.py:
```diff
@@ -107,19 +109,20 @@ def prepare(self, other_values, dataset_hash, user_hash, trans):
-                raise Exception("Dataset conversion failed for data parameter: %s" % self.name)
+            direct_match, target_ext, converted_dataset = data.find_conversion_destination(self.formats, converter_safe=True)
+            if direct_match:
+                elif target_ext and not converted_dataset:
```
```suggestion
                if target_ext and not converted_dataset:
```

bernt-matthias replied:
done

bernt-matthias committed:
one more mock hda conversion dest change

mvdbeek committed:
Default MockHDA to direct match

And change conversion_destination to `(False, None, None)` if there
shouldn't be a match.
